compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,292 @@
+      LDBG() << "FAILED - transfer size " << transferSizePerLane
+             << " does not match any target DMA size";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "transfer size does not match any target DMA size");
+    }
+    LDBG() << "Transfer size matches target DMA sizes";
+
+    // TODO: Handle indices properly - for now we skip the explicit indices
+    // check The lane parameter is always present but we handle it differently.
+
+    ArrayRef<int64_t> sourceShape = sourceType.getShape();
+    LDBG() << "Source rank: " << sourceShape.size();
+    if (sourceShape.size() < 2) {
+      LDBG() << "FAILED - source rank < 2";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "source must have at least 2 dimensions");
+    }
+
+    int64_t secondInnermostDimSize = sourceShape[sourceShape.size() - 2];
+    LDBG() << "  Source second innermost dimension size: "
+           << secondInnermostDimSize;
+
+    int64_t elementsPerTransfer = innermostDimSize / *subgroupSize;
+
+    auto transferType = VectorType::get({elementsPerTransfer}, elementType);
+
+    // Actually create the GatherToLDS ops to perform the transfer.
+    rewriter.setInsertionPoint(dmaOp);
+
+    TypedValue<IndexType> laneId = dmaOp.getLane();
+
+    Location loc = dmaOp.getLoc();
+    Value laneOffset = arith::MulIOp::create(
+        rewriter, loc, laneId,
+        arith::ConstantIndexOp::create(rewriter, loc, elementsPerTransfer));
+
+    Value zero = arith::ConstantIndexOp::create(rewriter, loc, 0);
+    for (int64_t i = 0; i < secondInnermostDimSize; ++i) {
+      Value iVal = arith::ConstantIndexOp::create(rewriter, loc, i);
+
+      SmallVector<Value> srcIndices;
+      for (int64_t dim = 0; dim < sourceType.getRank() - 2; ++dim) {
+        srcIndices.push_back(zero);
+      }
+      srcIndices.push_back(iVal);
+      srcIndices.push_back(laneOffset);
+
+      SmallVector<Value> dstIndices;
+      for (int64_t dim = 0; dim < destType.getRank() - 2; ++dim) {
+        dstIndices.push_back(zero);
+      }
+      dstIndices.push_back(iVal);
+      dstIndices.push_back(zero);
+
+      amdgpu::GatherToLDSOp::create(rewriter, loc, source, srcIndices, dest,
+                                    dstIndices, TypeAttr::get(transferType));
+    }
+
+    rewriter.eraseOp(dmaOp);
+    return success();
+  }
+
+private:
+  ArrayRef<int64_t> targetDmaSizes;
+};
+
+namespace {
+struct GPULowerCoalescedDMAToGlobalLoadsPass final
+    : impl::GPULowerCoalescedDMAToGlobalLoadsPassBase<
+          GPULowerCoalescedDMAToGlobalLoadsPass> {
+  void runOnOperation() override {
+    mlir::FunctionOpInterface funcOp = getOperation();
+
+    int count = 0;
+    funcOp.walk([&count](IREE::GPU::CoalescedGatherDMAOp op) {
+      LDBG() << "Found CoalescedGatherDMAOp: " << op;
+      ++count;
+    });
+    LDBG() << "Total CoalescedGatherDMAOps found: " << count;
+    if (count == 0) {
+      LDBG() << "No CoalescedGatherDMAOps found, exiting early";
+      return;
+    }
```

Comment:
IMO there's no point to traverse twice -- you can let the walk rewrite pattern rewrite drive run and not trigger on anything